Setting driver wide collection options for safe, fsync and timeout.

elricho opened the pull request:
Presently there is no way to set safe, fsync and timeout collection options for creates and updates.
This patch implements a driver field 'collection_options' => array(...)
The array can take 'safe' => n, 'fsync' => true/false, 'timeout' => n
See http://www.php.net/manual/en/mongocollection.save.php for more informaiton on field settings.

elricho committed:
Changes allow setting of driver wide colleciton options for safe, fsync & timeout.

elricho committed:
Fixed bug in insert introduced in previous commit.

YOMorales commented:
Perfect. I was about to do that myself (make update operations accept extra parameters other than 'multiple' => true). But thanks for getting ahead. I still changed something related to updates; see here: https://github.com/ichikaway/cakephp-mongodb/pull/31

elricho commented:
Cool. :) FYI - There was a bug in my insert, which is fixed now. :)

I was also thinking of extending it a little further with a mechanism that lets colection_options be set for one request only (I dont always need safe mode set for everything).  I'll model it on http://book.cakephp.org/view/1045/Creating-and-Destroying-Associations-on-the-Fly#!/view/1045/Creating-and-Destroying-Associations-on-the-Fly - Like binding the options for one request only.

YOMorales commented:
That is a better approach.

At first I thought of passing a third parameter in updateAll so that collection options could be set on demand like you say (something like: Mode->updateAll($fields, $conditions, $mongodb_options). But that would need some override of the Model class' updateAll method, and IMO this is something out of the 'scope' of this plugin.

But your suggestion is better. Something like binding.
It won't even need using a new method the way 'Binding Associations' does, but the simplest way is to set a class attribute. Something like:

$this->Model->collectionOptions = array('upsert' => true, 'safe' => true. etc.);
Then in Datasource...
if (!empty($this>collectionOptions)) {
$mongoCollectionObj->update($cond, $data, $this>collectionOptions);

AD7six commented:
It is not appropriate to use a db-wide options array on multiple collections - and on function calls that are expecting different options to be used.
- batchInsert has only safe and fsync
- insert has safe, fsync and timeout
- save has safe, fsync and timeout
- update has safe, fsync, timeout, upsert and multiple

It's likely that if you want to modify these settings, you want them different for specific collections and probably specific operations. By which I mean for example: inserts on [collection] are safe, but updating some counter/timestamp field is not a safe update. Therefore, while these changes may scratch your current itch it doesn't really solve "the problem". 

I'd suggest that where currently you reference `$this->config['collection_options']` it should effectively reference `$Model->mongoOptions[$methodname]` or similar. Unless it's possible to easily change these options at least on a per collection basis I don't see a benefit in changing hardcoded options for inflexible options.

elricho commented:
Hi Andy,

Thanks for the feedback and fair point. 

To expand on your suggestion, should the options in $Model->mongoOptions[$methodname] be set by $Model->setMongoOptions(array('save' => array('safe'), 'insert' => array('safe','fsync','upsert','timeout' => n), ...));
(ie getter/setter style) or just set it directly in the model like YOMorales' suggestion ?

Cheers
r

AD7six commented:
elricho - personally I'd go with simple - if (!empty($Model->...)) - but Ichi might have a different idea, and it's his code. However, if it works (by any means) it would be simple to make it work better before/after merging.

elricho committed:
Updated to use collectionOptions on a per collection basis rather than driver wide.

ichikaway commented:
sorry too late reply. 
I agree with using model property.
I will merge it,then change to use model property.
Now the patch fails test cases, I will fix it.

Thanks great idea and discussion.

redthor commented on models/datasources/mongodb_source.php:
```diff
@@ -710,33 +732,35 @@ public function update(&$Model, $fields = null, $values = null, $conditions = nu
 
 			$data = $this->setMongoUpdateOperator($Model, $data);
-
+			$params = array_merge($this->collectionOptions['update'], array("multiple" => true));
```
`$params = array_merge($this->collectionOptions['update'], array("multiple" => false));`

redthor commented on models/datasources/mongodb_source.php:
```diff
@@ -91,10 +91,30 @@ class MongodbSource extends DboSource {
+
+	public $collectionOptions = array(
+		'save' => array(),
+		'insert' => array(),
```
'insert' => array('safe' => true),

redthor commented:
Hi Ichi,
Not sure if you fixed the failed test cases...
I found that if I made these changes to the patch I could get all tests to pass:

line 112:
`'insert' => array('safe' => true),`

line 742:
`$params = array_merge($this->collectionOptions['update'], array("multiple" => false));`

Commit:
Library ported from using php-mongo extebsion to new php-mongodb extension.
Tests updated.
Composer added.

Commit:
Remove outdated directories.

Commit:
Remove outdated docs.

Commit:
Remove outdated Cakephp 1.3 samples.

Commit:
Remove debug statements.

Commit:
Update readme.

Commit:
Updated readme.

Commit:
Cleanup JSON for Packagist

Commit:
Remove uppercase chars from package name ! Doh, i always forget this.

Commit:
Adjust composer instructions for lowercase composer name.

Commit:
Add Travis CI

Commit:
Install php-mongodb extension for php 5.6
Enable extension in php 5.6 & 7.0
Add php 7.0 as test target (will this work with cakephp 2.8?)

Commit:
Use correct path to tests.

Commit:
Remove debug from tests.

Commit:
BugFix : Change MongoDate to new class MongoDB\BSON\UTCDateTime.

Commit:
BugFix : Replace MongoDate with MongoDB\BSON\UTCDateTime

Commit:
BugFIx : Use MongoDB::MONGODB_VERSION constant to get the driver version.

Commit:
BugFix : Update exceptions to new exception classes.
BugFix : Add try/catch blocks to all database calls.

Commit:
Have composer install deps.

Commit:
Ensure composer autoload.php executes before tests run.

Commit:
Fix previous blunder - doh !

Commit:
Add composer autoload to cakephp bootstrap.

Commit:
Update composer.json for correct phpunit version. (Cakephp 2.x needs phpunit 3.x).

Commit:
Remove reference to mysql.

Commit:
Cleanup better after tests.

Commit:
BugFix : testCreateConnectionName using incorrect datasource name.

Commit:
Fix datasource issues with $test database defaults.

Commit:
Added debug to tests - tracing a travis issue.

Commit:
Install Mongodb 3.2 (Travis uses 2.4).

Commit:
Don't drop system collections.
